src/execution/execute-afm.ts

```diff
@@ -1,137 +1,251 @@
+    : Execution.IExecutionResult {
+    return executionResultWrapper.executionResult;
 }
 
-// works only for one or two dimensions
-export function mergePageData(result: any, currentPage: any) {
-    const { paging, data, headerItems } = currentPage.executionResult;
-    const rowOffset = paging.offset[0];
-    if (result.executionResult.data[rowOffset]) { // appending columns to existing rows
-        for (let i = 0; i < data.length; i += 1) {
-            const columns = data[i] as any;
+function validateNumOfDimensions(numOfDimensions: number): void {
+    invariant(
+        numOfDimensions === 1 || numOfDimensions === 2,
+        `${numOfDimensions} dimensions are not allowed. Only 1 or 2 dimensions are supported.`
+    );
+}
 
-            result.executionResult.data[i + rowOffset].push(...columns);
-        }
-    } else { // appending new rows
-        result.executionResult.data.push(...data);
+function replaceLimitAndOffsetInUri(oldUri: string, limit: number[], offset: number[]): string {
+    const [uriPart, queryPart] = oldUri.split(/\?(.+)/);
+    const query = {
+        ...qs.parse(queryPart),
+        limit: limit.join(','),
+        offset: offset.join(',')
+    };
+
+    return uriPart + qs.stringify(query, { addQueryPrefix: true });
+}
+
+export function getNextOffset(limit: number[], offset: number[], total: number[]): number[] {
+    const numOfDimensions = total.length;
+    validateNumOfDimensions(numOfDimensions);
+
+    const defaultNextRowsOffset = offset[0] + limit[0];
+
+    if (numOfDimensions === 1) {
+        return [defaultNextRowsOffset];
+    }
+
+    const defaultNextColumnsOffset = offset[1] + limit[1];
+
+    const nextColumnsExist = offset[1] + limit[1] < total[1];
+
+    const nextRowsOffset = nextColumnsExist
+        ? offset[0] // stay in the same rows
+        : defaultNextRowsOffset; // go to the next rows
+
+    const nextColumnsOffset = nextColumnsExist
+        ? defaultNextColumnsOffset // next columns for the same rows
+        : 0; // start in the beginning of the next rows
+
+    return [nextRowsOffset, nextColumnsOffset];
+}
+
+export function getNextLimit(limit: number[], nextOffset: number[], total: number[]): number[] {
+    const numOfDimensions = total.length;
+    validateNumOfDimensions(numOfDimensions);
+
+    const getSingleNextLimit = (limit: number, nextOffset: number, total: number): number => nextOffset + limit > total
+        ? total - nextOffset
+        : limit;
+
+    // prevent set up lower limit than possible for 2nd dimension in the beginning of the next rows
+    if (
+        numOfDimensions === 2 &&
+        nextOffset[1] === 0 && // beginning of the next rows
+        limit[0] < total[1] // limit from 1st dimension should be used in 2nd dimension
+    ) {
+        return [
+            getSingleNextLimit(limit[0], nextOffset[0], total[0]),
+            limit[0]
+        ];
     }
```

Comment:
This can break, when non-square pages could be requested.
I suggest not refraining from using `DEFAULT_LIMIT`. The code would be easier. And since we dont have variable pageSize, its ok.

Reply:
@zbycz as above comment `prevent set up lower limit than possible for 2nd dimension in the beginning of the next rows` says, this is only for the specific case. For other cases, there is return statement at the end of the function which can't break when non-square pages are requested.